decisions/0002-back-service-elasticsearch-with-custom-implementation.md

```diff
@@ -0,0 +1,65 @@
+# 2. Back service elasticsearch with custom implementation
+
+Date: 2018-05-11
+
+## Status
+
+Accepted
+
+## Context
+
+The `service` command is has become a useful way for us to start, stop, restart,
+and check the status of the many service processes behind a commcare cluster.
+For the most part, these commands map very predictably to a start/stop/restart/status
+command on either [`service`](https://linux.die.net/man/8/service) or `supervisorctl`.
+
+Recently and in the past, we have found that `elasticsearch` does not stop or restart
+reliably. In particular, after running `service elasticsearch stop`,
+using `ps aux | grep elasticsearc[h]` may still show a running Elasticsearch
+process, even if `service` is reporting a stopped status. This then requires
+killing the rogue process in order for a `start` to bring up a genuinely new process.
+
+Before the addition of the `service` command, we already had the `restart-elasticsearch`
+command, which is conceptually (though currently not functionally) equivalent to
+`... service elasticsearch restart`. It does not address the above issue of rogue
+processes, but it does address some other issues, including a reminder to stop all pillows,
+as well as theoretically providing rolling (and thus no-downtime) restart,
+which we currently do not benefit from because we do not store data in duplicate.
+
+## Decision
+
+We will be changing `... service elasticsearch ...` from using simply using the default
+behavior to performing a more targeted set of actions that are guaranteed to achieve the
+stop/start/restart/status action, and we will also be deprecating the `restart-elasticsearch`
+entry point. In particular, `... service elasticsearch stop` will
+
+- stop pillows
+- guarantee the process has stopped and `kill -9` it after a period of time if it has not
+
+and `start` will
+
+- start pillows
+
+and `restart` will be functionally a `stop` followed by a `start`.
+
+Pros:
+- Someone with little context will be able to run a `... service elasticsearch ...`
+  and have it have the intended overall consequence
+- The tool will incorporate what is currently heard knowledge
+
+Cons:
+- A "power" user may think of the command as executing a very specific other command,
+  and may be taken aback by the tool trying to outsmart them.
```

Comment:
It's always frustrating to me when tools try to do this. I think that `service` is the exact command that runs `/etc/init.d/<service_name>` right? What do you think of having a different command like `cchq <env> safe_restart <service>`  that does these sorts of things? It makes it clear that it is a superset of whatever service management scheme we use now and in the future.

Reply:
I think this is OK given what @dannyroberts said in the next paragraph about the tool making it clear what's happening and also allowing the user to 'downgrade' to the simple version.

I also don't think having a separate command is a great idea since it would mean there are 2 commands for doing the same thing which requires a user to know which command to run in their context. It also means a user may not know about both and end up using the wrong one out of ignorance.